dmoj/cptbox/isolate.py

```diff
@@ -201,176 +202,290 @@ def __init__(self, read_fs, write_fs=None, writable=(1, 2)):
-            if error is not None:
-                return error
+            rel_file = self.get_rel_file(debugger, file_reg)
 
             # FIXME(tbrindus): defined here because FreeBSD 13 does not
             # implement AT_EMPTY_PATH, and 14 is not yet released (but does).
             AT_EMPTY_PATH = 0x1000
             # FIXME(tbrindus): we always follow symlinks, regardless of whether
             # AT_SYMLINK_NOFOLLOW is set. This may result in us denying files
             # we otherwise wouldn't have.
-            if kind == FilesystemSyscallKind.FSTAT and file == '' and debugger.uarg3 & AT_EMPTY_PATH:
+            if rel_file == '' and debugger.uarg3 & AT_EMPTY_PATH:
                 # If pathname is an empty string, operate on the file referred to
                 # by dirfd (which may have been obtained using the open(2) O_PATH
                 # flag). In this case, dirfd can refer to any type of file, not
                 # just a directory, and the behavior of fstatat() is similar to
                 # that of fstat(). If dirfd is AT_FDCWD, the call operates on the
                 # current working directory.
+                # We already allowed this one way or another, don't check again
                 return True
 
-            file, error = self._file_access_check(file, debugger, fs_jail(debugger), dirfd=debugger.arg0)
-            if not error:
-                return True
+            dirfd = self.get_reg_value(debugger, dir_reg)
+            full_path = self.get_full_path(debugger, rel_file, dirfd)
+            return self._raw_access_check(syscall, debugger, full_path, self.read_fs_jail)
+
+        return check
+
+    @staticmethod
+    def wrap_access_check_dec(syscall: str) -> Callable[[AccessChecker], HandlerCallback]:
+        def inner(func: AccessChecker) -> HandlerCallback:
+            return IsolateTracer.wrap_access_check(syscall, func)
+
+        return inner
+
+    @staticmethod
+    def wrap_access_check(syscall: str, access_check: AccessChecker) -> HandlerCallback:
+        def inner(debugger) -> bool:
+            try:
+                allowed = access_check(debugger)
+            except AccessFailure as failure:
+                failure.log_failure(syscall)
+                return failure.handle(debugger)
+            else:
+                if not allowed:
+                    return ACCESS_EACCES(debugger)
+                else:
+                    return True
+
+        return inner
+
+    DirFDGetter = Callable[[Debugger], int]
+
+    def _dirfd_getter_from_reg(self, reg: int) -> DirFDGetter:
+        def getter(debugger: Debugger) -> int:
+            return self.get_reg_value(debugger, reg)
+
+        return getter
 
-            log.debug('Denied access via syscall %s (error: %s): %s', syscall, error.error_name, file)
-            return error(debugger)
+    def _dirfd_getter_cwd(self, debugger: Debugger) -> int:
+        return AT_FDCWD
```

Comment:
How does this typecheck, returning an `int` instead of a `DirFDGetter`?

Reply:
Because it is a DirFDGetter, and those return ints.